Build: skip build based on commands' exit codes (#9649)
readthedocs/readthedocs.org: docs/user/build-customization.rst

         - git fetch --unshallow

> Skip build based on a condition
> ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Reviewer: ```suggestion
Skipping builds
~~~~~~~~~~~~~~~
```

Reviewer: While you are at it (SORRY!!!) - what about the automatic skipping, i.e. auto-canceling? https://blog.readthedocs.com/cancel-old-builds/

Reviewer: I mean, in this subsubsection or an adjacent section, you could quickly mention that builds are automatically skipped if canceled and how that works?

Author: I only added a quick example for now. I agree that we should document these two things in a different place. For now, I put this "feature" in this particular page because it makes sense under "Build customization" since it requires the user to change the normal flow.

However, as you may thought already, a specific feature page mentioning both of them would probably be a good addition in another PR.

Author: In fact, I think a good place for this content is https://docs.readthedocs.io/en/stable/builds.html

Reviewer: Moving it to the "Builds" page also sounds good, although a bit of how-to and explanation is still mixed.

I don't think you should feel blocked by the location of this content -- if you can develop the content, then we can move it later.

Author: I'm saying the code examples (how to) should live here, in `build-customization`. However, the introduction and description of the feature (explanation) should live in the `builds` page.

Author: I opened this PR that addresses these changes: https://github.com/readthedocs/readthedocs.org/pull/9717. Let's continue talking there, but I think it's 💯

There may be situations where you want to skip a build that was automatically triggered when someone on your team pushed to the repository.

> Skipping builds will allow you to speed up review times and also help us to reduce costs in our servers.

Reviewer: ```suggestion
Skipping builds will allow you to speed up review times and also help us reduce server costs and ultimately our environmental footprint.
```

> The following situations are good examples to skip a build:

Reviewer: ```suggestion
Consider the following scenarios:
```

* the build depends on an external situation that's not met yet
* there were no changes on the documentation files

In these scenarios, you can skip the build by executing a custom command that checks for that particular condition
and exits with code ``439`` to skip it, or ``0`` to continue building the documentation normally.
If any of the commands return this particular exit code,
Read the Docs will stop the build immediately,
mark it as "Cancelled",

> and communicate GitHub/GitLab that the build succeeded (green tick) so the pull request is in a mergeable state.

Reviewer: ```suggestion
and communicate to your Git platform (GitHub/GitLab/Bitbucket) that the build succeeded (green tick ✅) so the pull request is in a mergeable state.
```

Author: ```suggestion
and communicate to your Git platform (GitHub/GitLab) that the build succeeded (green tick ✅) so the pull request is in a mergeable state.
```

I updated the suggestion because we don't support PR builders on Bitbucket.

> .. code-block:: yaml

Reviewer: ```suggestion
Here is an example that exits the build when there are no changes to the `docs/` folder compared to the `origin/main` branch.

.. code-block:: yaml
```

Author: ```suggestion
Here is an example that exits the build when there are no changes to the ``docs/`` folder compared to the ``origin/main`` branch.

.. code-block:: yaml
```

   :caption: .readthedocs.yaml

   version: 2
   build:
     os: "ubuntu-22.04"
     tools:
       python: "3.11"
     jobs:
       post_checkout:
         # Skip build when there aren't changed in docs directory.
         # `--quiet` exits with a 1 when there **are** changes,
         # so we invert the logic with a !
         #
         # If there are no changes (exit 0) we force the command to return with 439.
         # This is a special exit code on Read the Docs that will cancel the build immediately.

>          - ! git diff --quiet origin/main -- docs/ && exit 439

Reviewer: @humitos as I've noted before in our threads about conditional builds.... doesn't this skip all builds on main itself?

Author: Hrm... Good point! I think you are right and this should only be run on pull requests 🤔

Author: I updated this code to use an `if` and an environment variable so it only runs on pull requests.

Reviewer: I think the new solution looks good :+1: 

It might benefit from one of the multi-line yaml syntax?

```
key: |
   value
   blah
   blah
```

(this syntax preserves the newlines btw)

Reviewer: Circle CI uses this syntax a lot :arrow_up:

Author: He, I was updating this at the same time you were writing this comment. See [readthedocs/readthedocs.org@`1d4ae24` (#9717)](https://github.com/readthedocs/readthedocs.org/pull/9717/commits/1d4ae249e740e4ee7492f6c54cdb83e23d093f83)

Reviewer: We are truly working in the same timezone :smile:

In these scenarios, you can skip the build by executing a custom command that checks for that particular condition
and exits with code ``439`` to skip it, or ``0`` to continue building the documentation normally.
If any of the commands return this particular exit code,
Read the Docs will stop the build immediately,
mark it as "Cancelled",
and communicate to your Git platform (GitHub/GitLab) that the build succeeded (green tick ✅) so the pull request is in a mergeable state.

Here is an example that skip build from pull requests when there are no changes to the ``docs/`` folder compared to the ``origin/main`` branch.

.. code-block:: yaml
   :caption: .readthedocs.yaml

   version: 2
   build:
     os: "ubuntu-22.04"
     tools:
       python: "3.11"
     jobs:
       post_checkout:
         # Skip building pull requests when there aren't changed in the docs directory.
         # `--quiet` exits with a 1 when there **are** changes,
         # so we invert the logic with a !
         #
         # If there are no changes (exit 0) we force the command to return with 439.
         # This is a special exit code on Read the Docs that will cancel the build immediately.

>          - if [ $READTHEDOCS_VERSION_TYPE = "external" ]; then ! git diff --quiet origin/main -- docs/ && exit 439; fi

Reviewer 2: Hello. Would be nice to have example to show to do implement "skip RTD build if there is a [ci skip], [skip ci], [skip rtd], or [rtd skip]" in the last commit message. Thanks!

Reviewer 2: Thanks! I am excited to try it out when it is available. 😸

Author: You can already test this 😉 , but keep reading...

... the only difference with this PR is that (when merged and deployed) it:

* will show a nicer error message instead of saying that your build is broken
* will send "Success" (green tick ✅ ) to GitHub when skipping the build instead "Failed" (red cross ❌ )

The links that I've shared are from production already 😎

         - git fetch --unshallow

Cancel build based on a condition
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> When a command exits with code ``183``,

Reviewer 3: We should probably explain why we chose 183, since it definitely seems somewhat arbitrary :)

Author: I added a small note for this.

.. code-block:: yaml
   :caption: .readthedocs.yaml

   version: 2
   build:
     os: "ubuntu-22.04"
     tools:
       python: "3.11"
     jobs:
       post_checkout:
         # Cancel building pull requests when there aren't changed in the docs directory.
         # `--quiet` exits with a 1 when there **are** changes,
         # so we invert the logic with a !
         #
         # If there are no changes (exit 0) we force the command to return with 183.
         # This is a special exit code on Read the Docs that will cancel the build immediately.
         - |
           if [ $READTHEDOCS_VERSION_TYPE = "external" ];
           then
             ! git diff --quiet origin/main -- docs/ && exit 183;
           fi

This other example shows how to cancel a build if the commit message contains ``skip ci`` on it:

.. code-block:: yaml
   :caption: .readthedocs.yaml

   version: 2
   build:
     os: "ubuntu-22.04"
     tools:
       python: "3.11"
     jobs:
       post_checkout:
         # Use `git log` to check if the latest commit contains "skip ci",
         # in that case exit the command with 183 to cancel the build
         - |

>            case `git --no-pager log --pretty="tformat:%s" -1`
>            in *"skip ci"*)
>              exit 183;;
>            *);;
>            esac

Reviewer 4: `case` has some rather gnarly syntax to it, is there any reason we're recommending this over `grep`?

```suggestion
           (git --no-pager log --pretty="tformat:%s" -1` | grep -q "skip ci") && exit 183
```

(Not tested against this PR, but should be close)

Author: You are the bash ninja. So, I'll let you decide what's better 👍

Reviewer 4: Hah well I also haven't followed this part too closely, so you might have more reasoning for `case`. But if there is no strong preference, `grep` is more commonly used, so might be my suggestion.

My example was a little off, but this seems to be it:

```
(git --no-pager log --pretty="tformat:%s" -1 | grep -viq "skip ci") || exit 183
```

Note, the double inversion of `grep -v` and `||` to give either exit `0` on a miss or `183` on a hit of the search term:

```
(community) community/ main* 💀  git log -n 1                                                                                   
02dd7b34d (HEAD -> main, origin/main, origin/HEAD) Proxito: don't depend on attributes injected in the request (#9711) <stsewd@proton.me> 5 hours ago
(community) community/ main* 💀  /bin/sh -c '(git --no-pager log --pretty="tformat:%s" -1 | grep -viq "miss") || exit 183'
(community) community/ main* 💀  echo $?
0
(community) community/ main* 💀  /bin/sh -c '(git --no-pager log --pretty="tformat:%s" -1 | grep -viq "proxito") || exit 183'
(community) community/ main* 🔥  echo $?
183
```

Reviewer 2: I don't think that grep example can do multiple directives? What if I want to match "skip ci" or "ci skip"?

Reviewer 3: grep can do regex: `grep -E 'ci skip|skip ci'`

Author: I updated this example to use `grep`, 👍🏼

Reviewer 2: Hmm. My shell-fu is weak and I cannot get this to work. It doesn't skip.

```
(git --no-pager log --pretty="tformat:%s -- %b" -1 | grep -vqE "skip rtd|rtd skip") || exit 183
```

.. code-block:: yaml
   :caption: .readthedocs.yaml

   version: 2
   build:
     os: "ubuntu-22.04"
     tools:
       python: "3.11"
     jobs:
       post_checkout:
         # Skip building pull requests when there aren't changed in the docs directory.
         # `--quiet` exits with a 1 when there **are** changes,
         # so we invert the logic with a !
         #
         # If there are no changes (exit 0) we force the command to return with 439.
         # This is a special exit code on Read the Docs that will cancel the build immediately.
         - if [ $READTHEDOCS_VERSION_TYPE = "external" ]; then ! git diff --quiet origin/main -- docs/ && exit 439; fi

This other example shows how to skip a build if the commit message contains ``skip ci`` on it:

.. code-block:: yaml
   :caption: .readthedocs.yaml

   version: 2
   build:
     os: "ubuntu-22.04"
     tools:
       python: "3.11"
     jobs:
       post_checkout:
         # Use `git log` to check if the latest commit contains "skip ci",
         # in that case exit the command with 439 to skip the build

>          - case `git --no-pager log --pretty="tformat:%s" -1` in *"skip ci"*) exit 439;; *);; esac

Reviewer 2: Hmm. I implemented this:

https://github.com/spacetelescope/jdaviz/commit/b15517753139916ed5598ef1871fa955f12da8d4

And then I tried using the directive in https://github.com/spacetelescope/jdaviz/pull/1829 which created https://readthedocs.org/projects/jdaviz/builds/18586511/ that did not get cancelled.

What did I do wrong?

Author: That's because that `git log` command only outputs the _title_ of the commit, without the description. Checking over the description as well, may require a different `--pretty=` format argument.

Author: In that case, you may want something like:

```
$ git --no-pager log --pretty="tformat:%s -- %b" -1
Use multi-line bash examples -- They are easier to read
```

Author: note the `%s` _and_ `%b`

I found this at https://git-scm.com/docs/pretty-formats

Reviewer 2: I think that works. Thanks! https://readthedocs.org/projects/jdaviz/builds/18586938/

However, would be nice if such checks is done before cancelling a previous job from the same branch. Consider this use case:

1. Commit A pushed with real changes. It is still running the RTD build.
2. Commit B is trivial and has skip directive in its commit message. It is pushed while the run for Commit A is still not finished.

Now, RTD will cancel the run for Commit A and then fail the build for Commit B (because of the skip). Commit A should keep running if Commit B is skipped.

Author: Gotcha! This would be pretty hard to implement with the current code and structure.

If you want this to be considered for a next iteration of this feature, please open a new issue explaining these details 👍

You have shared great feedback already! 💯

Author: Actually, Read the Docs know nothing about the reason of the build being skipped, so I don't think it will be able to make these decisions 🤷‍♂️

Reviewer 2: Well, in other CI services, skip directives (e.g., `[ci skip]` in commit message) are defined internally, so I guess that would make things easier. For example, Actions doesn't even start when it detects the skip (for better or worse), hence avoid cancelling the last job even if auto-cancel is defined in its workflow.

So, should I still bother to open a new issue? Or maybe at least document the canceling behavior in the doc?

Author: The cancelling behavior is already documented in another PR (I'm on mobile, but you can quickly find it)

The "skip ci" behaviour from other CIs is exactly what you need for this use case. However, this feature of using exit codes to get the outcome is not exactly that feature, but am approximation.

It's not integrated into the CI itself, but using another feature (the one implemented on this PR) to have a similar behavior.

Because it's not implemented at the core of the CI we can't handle these particular cases.

We should handle it in a different way, by exploring the commit message from the webhook itself and decide to no trigger the build at all.

What I'm describing is a different feature from what this PR implements. A valid feature, tho.

Author: BTW, this is the commit that documents "When to cancel builds" and explains how it works: [`111ee35` (#9649)](https://github.com/readthedocs/readthedocs.org/pull/9649/commits/111ee350847d5ce3857a24078a7b21692bd950fd)

Reviewer: @humitos let's say that a build is automatically cancelled and superseded by a build that's then also cancelled.

Is there a button on auto-cancelled builds that can force them to run, i.e. this one?

![image](https://user-images.githubusercontent.com/374612/201088670-74c08773-635d-4e1e-9f98-18c5566d0881.png)

(sorry, I couldn't quickly find an auto-cancelled build)

Author: Yeah, we could change the logic behind that button if we want to.

https://github.com/readthedocs/readthedocs.org/blob/eeb62b71afde2d865cbb59145ebfd7b0500c799c/readthedocs/templates/builds/build_detail.html#L80-L82

However, if I'm reading that comment correctly, I think it should be present in the case that @pllim was describing.

Reviewer 2: Not autocancelling is better but I guess a button to force rebuild is better than nothing. Thanks!

So, I should have been using exit code 183 and not 439, even after this PR is merged?

Author: The exit code is the same. POSIX automatically calculate the module of integers bigger than 255

Reviewer 2: I guess it works. 😸 https://readthedocs.org/projects/jdaviz/builds/18597605/

         - git fetch --unshallow

Cancel build based on a condition
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

When a command exits with code ``183``,
Read the Docs will cancel the build immediately.
You can use this approach to cancel builds that you don't want to complete based on some conditional logic.

.. note:: Why 183 was chosen for the exit code?

   It's the word "skip" encoded in ASCII.
   Then it's taken the 256 modulo of it because
   `the Unix implementation does this automatically <https://tldp.org/LDP/abs/html/exitcodes.html>`_
   for exit codes greater than 255.

   .. code-block:: python

      >>> sum(list('skip'.encode('ascii')))

>       439

Reviewer 2: Belated comment: Do you still need this explanation now that you no longer use 439 ?

Author: It explains where the number 183 comes from. Isn't it clear with this note? I'm happy to update it
